Create last_update_export.py

josephoaks opened the pull request:
This script is designed for use in a "disconnect" environment. The host server is to do daily exports based on last build of a channel. This script looks at the list of channels, gets the ChannelLastBuildByID and checks if the date is the current date or prior and then exports that channel.

Logging is provided as well to show which channels were updated in that export. 

The target server can then rsync the `/mnt/export` directory and run another script to execute the inter-server-sync import of the exported data.

josephoaks committed:
Create last_update_export.py

This script is designed for use in a "disconnect" environment. The host server is to do daily exports based on last build of a channel. This script looks at the list of channels, gets the ChannelLastBuildByID and checks if the date is the current date or prior and then exports that channel.

Logging is provided as well to show which channels were updated in that export. 

The target server can then rsync the `/mnt/export` directory and run another script to execute the inter-server-sync import of the exported data.

meaksh commented on uyuni-tools/last_update_export.py:
```diff
@@ -0,0 +1,42 @@
+#! /usr/bin/python3
+
```
Could you please ask some documentation about this script as header? Including description, instructions and intended usage?

josephoaks replied:
Added script header

meaksh commented on uyuni-tools/last_update_export.py:
```diff
@@ -0,0 +1,42 @@
+
+today = datetime.date.today()
+yesterday = today - datetime.timedelta(days=1)
+log_file_path = f"/mnt/logs/{today}-daily_export.log"
+output_dir = "/mnt/export/updates"
```
The script is assuming these paths are existing but this is not usually the case.

Could you make these configurable as other parameters? Then it would be nice that the script will check first if the paths exist, to prevent problems.

meaksh commented on uyuni-tools/last_update_export.py:
```diff
@@ -0,0 +1,42 @@
+log_file_path = f"/mnt/logs/{today}-daily_export.log"
+output_dir = "/mnt/export/updates"
+
+subprocess.run(f"rm -rf {output_dir}/*", shell=True, check=True)
```
Maybe better to use to use `shutil.rmtree`, to avoid calling a subprocess.

meaksh commented on uyuni-tools/last_update_export.py:
```diff
@@ -0,0 +1,42 @@
+key = client.auth.login(MANAGER_LOGIN, MANAGER_PASSWORD)
+
+today = datetime.date.today()
+yesterday = today - datetime.timedelta(days=1)
```
Could this interval be configurable?

meaksh reviewed (changes requested):
Thanks for this contribution!

I've added few comments that I hope it helps to get the script in shape to get merged.

josephoaks committed:
Update last_update_export.py

Updates from the source since uploading to uyuni-tools, multiple iterations to streamline have occured.

Updates include fixes from comments and further use, which include...
1) script header
2) better path management and configurable parameters
3) use of shutil.rmtree avoiding extra calls to subprocess
4) configurable days for export
5) configurable user/group
6) added the use of mgr-sycn credentials file instead of using username/password in the script.

meaksh commented on uyuni-tools/last_update_export.py:
```diff
@@ -0,0 +1,103 @@
+    client = ServerProxy(MANAGER_URL, context=context)
+    return client, client.auth.login(MANAGER_LOGIN, MANAGER_PASSWORD)
+
+# Configuration Variables
+base_dir = "/mnt"  # Define base directory where the exports will be.
+output_dir = os.path.join(base_dir, "export/updates")
+log_dir = os.path.join(base_dir, "logs")
+today = datetime.date.today()
+target_date = today - datetime.timedelta(days=1)  # Define the number of days back for export, 1 day by default
+rsync_user = "rsyncuser"  # Define rsync user
+rsync_group = "users"     # Define rsync group
```
OTOH these ones I would move them to the top of the file (below imports statements and before functions definition), as user is expected to modify them.

I'd also suggest to use uppercase notation (constants) for those.

meaksh reviewed (commented):
Thanks for the changes! This is looking better now! Still I added some comments after your latest changes, maybe you can have a look.

meaksh commented on uyuni-tools/last_update_export.py:
```diff
@@ -0,0 +1,103 @@
+    config = configparser.ConfigParser()
+    with open(config_path, 'r') as f:
+        config.read_string('[DEFAULT]\n' + f.read())
+    MANAGER_LOGIN = config.get('DEFAULT', 'mgrsync.user')
+    MANAGER_PASSWORD = config.get('DEFAULT', 'mgrsync.password')
+    SUMA_FQDN = socket.getfqdn()
+    MANAGER_URL = f"https://{SUMA_FQDN}/rpc/api"
+    context = ssl.create_default_context()
+    client = ServerProxy(MANAGER_URL, context=context)
+    return client, client.auth.login(MANAGER_LOGIN, MANAGER_PASSWORD)
```
I would suggest to not use the uppercase notation here, as these values are not hardcoded anymore, and contants are usually defined at the top level of the file.

Maybe just go with something like:
```suggestion
    manager_login = config.get('DEFAULT', 'mgrsync.user')
    manager_password = config.get('DEFAULT', 'mgrsync.password')
    suma_fqdn = socket.getfqdn()
    manager_url = f"https://{suma_fqdn}/rpc/api"
    context = ssl.create_default_context()
    client = ServerProxy(manager_url, context=context)
    return client, client.auth.login(manager_login, manager_password)
```

Majus457 replied:
modified.

josephoaks committed:
Update last_update_export.py

updated the constants and variables, and changed the order they appear in the script

josephoaks committed:
Update last_update_export.py

caught the datetime.date.today()

meaksh commented on uyuni-tools/last_update_export.py:
```diff
@@ -0,0 +1,103 @@
+    build_date, channel_label = client.channel.software.getChannelLastBuildById(key, channel["id"]).split()[0], channel["label"]
+    build_date = datetime.datetime.strptime(build_date, "%Y-%m-%d").date()
+
+    if build_date in [TODAY, TARGET_DATE]:
```
I think this would only work fine if the interval is 1 day, but since the interval is configurable, it would not behave fine when interval is, i.a. 7 days. The condition would only check if `build_date` is equal to `TODAY` or `TARGET_DATE` but not in between.

I think you could use something like:

```suggestion
    if TARGET_DATE <= build_date <= TODAY:
```

meaksh commented on uyuni-tools/last_update_export.py:
```diff
@@ -0,0 +1,103 @@
+        options = ' '.join([f"--{opt}='{val}'" for opt, val in options_dict.items()])
+        command = f"inter-server-sync export --channels='{channel_label}' {options}"
+        with open(log_file_path, "a") as log_file:
+            subprocess.run(command, shell=True, stdout=log_file, stderr=subprocess.STDOUT, check=True)
```
In general I would recommend to use `subprocess.run` without `shell=True` (if possible), as it would prevent potential shell injections. as you do already here https://github.com/uyuni-project/contrib/pull/23/files#diff-828d5f5bff6d9517a822e3e503fa62753bbc631f61c1ffccc88c601bf1111ae2R103

josephoaks committed:
Update uyuni-tools/last_update_export.py

Co-authored-by: Pablo Suárez Hernández <psuarezhernandez@suse.com>

josephoaks committed:
Update last_update_export.py

removed the shell=True from the subprocess.run command, missed that from the last commit after updating the command

josephoaks committed:
Fixed the shebang line

Fixed the shebang line

meaksh commented on uyuni-tools/last_update_export.py:
```diff
@@ -0,0 +1,103 @@
+        options = ' '.join([f"--{opt}='{val}'" for opt, val in options_dict.items()])
+        command = f"inter-server-sync export --channels='{channel_label}' {options}"
+        with open(log_file_path, "a") as log_file:
+            subprocess.run(command, stdout=log_file, stderr=subprocess.STDOUT, check=True)
```
Hmmm, I think this won't work at the moment. If you remove the `shell=True`, then you must provide the `command` as a list (not as a string).

josephoaks replied:
resolved, thanks for pointing this out, when I ran my test before last commit it didn't throw up any errors, but did not work so I assumed it was good. Verified this is working on several servers before committed.

josephoaks committed:
Update last_update_export.py

fixed the command execution

josephoaks committed:
Update last_update_export.py script header

updated script header to be better formatted and easier to read.

josephoaks committed:
Create initial_export.py

This file is part of the last_update_export.py script put forward. 
This file is for the initial export using the ISSv2 used in a disconnected or air-gapped environment, to do a full export instead of an incremental export as the last_update_export.py, it can also be used to add a new product channel to an air-gapped system.

mbrookhuis commented:
Sorry for the late reaction. 
Could you review the section for getting the user and password. The script should not been run inside the container but rather outside. Maybe use the configsm.yaml.

josephoaks commented:
These scripts were not designed for the containerized version of SUMA. I will have to work on this.
